# Introduction to Numpy

More info: [http://wiki.scipy.org/Tentative_NumPy_Tutorial](http://wiki.scipy.org/Tentative_NumPy_Tutorial)

## Numpy Overview

* Why Python for Data? Numpy brings *decades* of C math into Python!
* Numpy provides a wrapper for extensive C/C++/Fortran codebases, used for data analysis functionality
* NDAarray allows easy vectorized math and broadcasting (i.e. functions for vector elements of different shapes)

In [1]:
import numpy as np

### Creating ndarrays

An array object represents a multidimensional, homogeneous array of fixed-size items. 

In [2]:
# Creating arrays
a = np.zeros((3))
b = np.ones((2,3))
c = np.random.randint(1,10,(2,3,4))
d = np.arange(0,11,1)

In [3]:
# Note the way each array is printed:
print(a)
print(b)
print(c)
print(d)

[0. 0. 0.]
[[1. 1. 1.]
 [1. 1. 1.]]
[[[9 2 1 9]
  [5 6 1 3]
  [1 7 2 1]]

 [[7 4 7 3]
  [1 1 7 7]
  [8 4 2 9]]]
[ 0  1  2  3  4  5  6  7  8  9 10]


### Arithmetic in arrays is element wise

In [4]:
a = np.array( [20,30,40,50] )
b = np.arange(4)
print(a)
print(b)

[20 30 40 50]
[0 1 2 3]


In [5]:
c = a - b
c

array([20, 29, 38, 47])

In [6]:
b**2

array([0, 1, 4, 9])

### Indexing, Slicing and Iterating

In [7]:
# one-dimensional arrays work like lists:
a = np.arange(10)**2
a

array([ 0,  1,  4,  9, 16, 25, 36, 49, 64, 81])

In [8]:
a[2:5]

array([ 4,  9, 16])

In [9]:
# Multidimensional arrays use tuples with commas for indexing
# with (row,column) conventions beginning, as always in Python, from 0

In [10]:
b = np.random.randint(1,100,(4,4))

In [11]:
b

array([[ 8, 50, 72, 75],
       [77, 16, 96, 62],
       [36, 46, 10, 30],
       [90, 72, 38, 85]])

In [12]:
# Guess the output
print(b[2,3])
print(b[0,0])

30
8


In [13]:
print(b[0:3,1])
print(b[:,1:3])

[50 16 46]
[[50 72]
 [16 96]
 [46 10]
 [72 38]]


In [14]:
b[1:3,:]

array([[77, 16, 96, 62],
       [36, 46, 10, 30]])

# Introduction to Pandas

## Pandas Overview

In [1]:
import pandas as pd
import numpy as np

###Import a dataset 

Use shift + tab to invoke the parameter intellisense 

In [2]:
passengers = pd.read_csv("data/air_passengers/air-passenger-arrivals-by-country-of-embarkation.csv")

### View the first five rows

In [17]:
passengers.head()

month                           level_1          level_2      level_3  \
0  1961-01  Number Of Air Passenger Arrivals  South East Asia     Malaysia   
1  1961-01  Number Of Air Passenger Arrivals  South East Asia    Indonesia   
2  1961-01  Number Of Air Passenger Arrivals  South East Asia     Thailand   
3  1961-01  Number Of Air Passenger Arrivals  South East Asia  Philippines   
4  1961-01  Number Of Air Passenger Arrivals  South East Asia      Vietnam   

  value  
0    na  
1  1687  
2  1172  
3   139  
4    na

### Bottom seven rows

In [18]:
passengers.tail(7)

month                           level_1          level_2  \
7418  2017-03  Number Of Air Passenger Arrivals  South East Asia   
7419  2017-03  Number Of Air Passenger Arrivals  North East Asia   
7420  2017-03  Number Of Air Passenger Arrivals  North East Asia   
7421  2017-03  Number Of Air Passenger Arrivals  North East Asia   
7422  2017-03  Number Of Air Passenger Arrivals           Europe   
7423  2017-03  Number Of Air Passenger Arrivals           Europe   
7424  2017-03  Number Of Air Passenger Arrivals           Europe   

             level_3   value  
7418         Vietnam  102567  
7419           China  243818  
7420       Hong Kong  146329  
7421           Japan  115925  
7422  United Kingdom   50101  
7423          France   16911  
7424         Germany   34072

### Length of dataset

In [19]:
len(passengers)

7425

In [20]:
passengers.shape

(7425, 5)

### Count for each variable in the dataset

In [21]:
passengers.count()

month      7425
level_1    7425
level_2    7425
level_3    7425
value      7425
dtype: int64

### Statistics for the dataset

In [22]:
passengers.describe()

month                           level_1          level_2    level_3  \
count      7425                              7425             7425       7425   
unique      675                                 1                3         11   
top     1983-08  Number Of Air Passenger Arrivals  South East Asia  Indonesia   
freq         11                              7425             3375        675   

       value  
count   7425  
unique  5841  
top       na  
freq    1368

### Datatypes for each column, automatically inferred by Pandas

In [23]:
passengers.dtypes

month      object
level_1    object
level_2    object
level_3    object
value      object
dtype: object

Value should be numeric, but null values are specified as strings "na" or "-"

### Re-importing the data, specifying null values and column names

In [4]:
passengers = pd.read_csv("data/air_passengers/air-passenger-arrivals-by-country-of-embarkation.csv", na_values=["na", "-"],
                        names=["year_month", "total", "region", "country", "value"], header=0)

### Describe by default shows stats for numeric columns

In [36]:
passengers.describe()

value
count    6056.000000
mean    55459.162979
std     61370.732088
min        19.000000
25%     10647.250000
50%     32011.500000
75%     82153.000000
max    372597.000000

### Explicitly describing all columns

In [37]:
passengers.describe(include='all')

year_month                             total           region  \
count        7425                              7425             7425   
unique        675                                 1                3   
top       1983-08  Number Of Air Passenger Arrivals  South East Asia   
freq           11                              7425             3375   
mean          NaN                               NaN              NaN   
std           NaN                               NaN              NaN   
min           NaN                               NaN              NaN   
25%           NaN                               NaN              NaN   
50%           NaN                               NaN              NaN   
75%           NaN                               NaN              NaN   
max           NaN                               NaN              NaN   

          country          value  
count        7425    6056.000000  
unique         11            NaN  
top     Indonesia            NaN  
freq          675            NaN  
mean          NaN   55459.162979  
std           NaN   61370.732088  
min           NaN      19.000000  
25%           NaN   10647.250000  
50%           NaN   32011.500000  
75%           NaN   82153.000000  
max           NaN  372597.000000

In [5]:
# Use value_counts() to count the frequency of unique entities under the specific column 
passengers["country"].value_counts()

Malaysia          675
Philippines       675
Germany           675
Indonesia         675
Hong Kong         675
United Kingdom    675
Thailand          675
Vietnam           675
China             675
France            675
Japan             675
Name: country, dtype: int64

### Total visitor arrivals by country

In [6]:
# Perform aggregation by country and value columns
passengers.groupby("country")["value"].sum()

country
China             29866493.0
France             4462481.0
Germany            8438001.0
Hong Kong         35468033.0
Indonesia         63995595.0
Japan             30487564.0
Malaysia          64924645.0
Philippines       18466360.0
Thailand          50736256.0
United Kingdom    17950341.0
Vietnam           11064922.0
Name: value, dtype: float64

### Sort the values by descending

In [40]:
passengers.groupby("country")["value"].sum().sort_values(ascending=False)

country
Malaysia          64924645.0
Indonesia         63995595.0
Thailand          50736256.0
Hong Kong         35468033.0
Japan             30487564.0
China             29866493.0
Philippines       18466360.0
United Kingdom    17950341.0
Vietnam           11064922.0
Germany            8438001.0
France             4462481.0
Name: value, dtype: float64

### Filter results by year-month = '2017-01'

In [41]:
passengers[passengers["year_month"] == "2017-01"]

year_month                             total           region  \
7392    2017-01  Number Of Air Passenger Arrivals  South East Asia   
7393    2017-01  Number Of Air Passenger Arrivals  South East Asia   
7394    2017-01  Number Of Air Passenger Arrivals  South East Asia   
7395    2017-01  Number Of Air Passenger Arrivals  South East Asia   
7396    2017-01  Number Of Air Passenger Arrivals  South East Asia   
7397    2017-01  Number Of Air Passenger Arrivals  North East Asia   
7398    2017-01  Number Of Air Passenger Arrivals  North East Asia   
7399    2017-01  Number Of Air Passenger Arrivals  North East Asia   
7400    2017-01  Number Of Air Passenger Arrivals           Europe   
7401    2017-01  Number Of Air Passenger Arrivals           Europe   
7402    2017-01  Number Of Air Passenger Arrivals           Europe   

             country     value  
7392        Malaysia  259314.0  
7393       Indonesia  308766.0  
7394        Thailand  244080.0  
7395     Philippines  139796.0  
7396         Vietnam   98670.0  
7397           China  296317.0  
7398       Hong Kong  161089.0  
7399           Japan  109393.0  
7400  United Kingdom   52431.0  
7401          France   18160.0  
7402         Germany   32906.0

In [44]:
# Recently, Pandas also introduced a SQL method: query, where users can query the column using conditions 
passengers.query('year_month == "2017-01"')

year_month                             total           region  \
7392    2017-01  Number Of Air Passenger Arrivals  South East Asia   
7393    2017-01  Number Of Air Passenger Arrivals  South East Asia   
7394    2017-01  Number Of Air Passenger Arrivals  South East Asia   
7395    2017-01  Number Of Air Passenger Arrivals  South East Asia   
7396    2017-01  Number Of Air Passenger Arrivals  South East Asia   
7397    2017-01  Number Of Air Passenger Arrivals  North East Asia   
7398    2017-01  Number Of Air Passenger Arrivals  North East Asia   
7399    2017-01  Number Of Air Passenger Arrivals  North East Asia   
7400    2017-01  Number Of Air Passenger Arrivals           Europe   
7401    2017-01  Number Of Air Passenger Arrivals           Europe   
7402    2017-01  Number Of Air Passenger Arrivals           Europe   

             country     value  
7392        Malaysia  259314.0  
7393       Indonesia  308766.0  
7394        Thailand  244080.0  
7395     Philippines  139796.0  
7396         Vietnam   98670.0  
7397           China  296317.0  
7398       Hong Kong  161089.0  
7399           Japan  109393.0  
7400  United Kingdom   52431.0  
7401          France   18160.0  
7402         Germany   32906.0

### and sort by value

In [46]:
passengers_201701 = passengers[passengers["year_month"] == "2017-01"]

In [47]:
passengers_201701.sort_values(by="value", ascending=False)

year_month                             total           region  \
7393    2017-01  Number Of Air Passenger Arrivals  South East Asia   
7397    2017-01  Number Of Air Passenger Arrivals  North East Asia   
7392    2017-01  Number Of Air Passenger Arrivals  South East Asia   
7394    2017-01  Number Of Air Passenger Arrivals  South East Asia   
7398    2017-01  Number Of Air Passenger Arrivals  North East Asia   
7395    2017-01  Number Of Air Passenger Arrivals  South East Asia   
7399    2017-01  Number Of Air Passenger Arrivals  North East Asia   
7396    2017-01  Number Of Air Passenger Arrivals  South East Asia   
7400    2017-01  Number Of Air Passenger Arrivals           Europe   
7402    2017-01  Number Of Air Passenger Arrivals           Europe   
7401    2017-01  Number Of Air Passenger Arrivals           Europe   

             country     value  
7393       Indonesia  308766.0  
7397           China  296317.0  
7392        Malaysia  259314.0  
7394        Thailand  244080.0  
7398       Hong Kong  161089.0  
7395     Philippines  139796.0  
7399           Japan  109393.0  
7396         Vietnam   98670.0  
7400  United Kingdom   52431.0  
7402         Germany   32906.0  
7401          France   18160.0

### What are the top five highest total arrivals by year-month?

In [48]:
passengers.groupby("year_month").sum().sort_values(by="value", ascending=False).head(5)

value
year_month           
2016-12     1813414.0
2016-07     1747537.0
2017-01     1720922.0
2017-03     1712386.0
2015-12     1710942.0

### Which countries have the most number of months with arrivals > 100000

In [49]:
passengers[passengers["value"] > 100000]["country"].value_counts()

Malaysia       311
Indonesia      269
Thailand       225
China          144
Hong Kong      108
Japan           51
Philippines     45
Vietnam          5
Name: country, dtype: int64

## Adding columns for year and month

In [51]:
passengers["year"] = passengers["year_month"].apply(lambda x: int(x.split("-")[0]))

In [52]:
passengers["month"] = passengers["year_month"].apply(lambda x: int(x.split("-")[1]))

In [9]:
passengers["month2"] = passengers["year_month"].str.split("-").str[1]

In [10]:
passengers.head()

year_month                             total           region      country  \
0    1961-01  Number Of Air Passenger Arrivals  South East Asia     Malaysia   
1    1961-01  Number Of Air Passenger Arrivals  South East Asia    Indonesia   
2    1961-01  Number Of Air Passenger Arrivals  South East Asia     Thailand   
3    1961-01  Number Of Air Passenger Arrivals  South East Asia  Philippines   
4    1961-01  Number Of Air Passenger Arrivals  South East Asia      Vietnam   

    value month2  
0     NaN     01  
1  1687.0     01  
2  1172.0     01  
3   139.0     01  
4     NaN     01

### Average number of passenger arrivals by month

In [54]:
passengers.groupby("month").mean()

value    year
month                      
1      54484.202729  1989.0
2      52060.263158  1989.0
3      56602.124756  1989.0
4      51791.027888  1988.5
5      50962.653386  1988.5
6      57080.749004  1988.5
7      57530.515936  1988.5
8      57577.836653  1988.5
9      53150.388446  1988.5
10     55514.668663  1988.5
11     55295.302789  1988.5
12     63531.129482  1988.5

### Sort descending by value

In [55]:
passengers.groupby("month")["value"].mean().sort_values(ascending=False)

month
12    63531.129482
8     57577.836653
7     57530.515936
6     57080.749004
3     56602.124756
10    55514.668663
11    55295.302789
1     54484.202729
9     53150.388446
2     52060.263158
4     51791.027888
5     50962.653386
Name: value, dtype: float64

# Next Steps

**Recommended Resources**

Name | Description
--- | ---
[Official Pandas Tutorials](http://pandas.pydata.org/pandas-docs/stable/10min.html) | Wes & Company's selection of tutorials and lectures
[Julia Evans Pandas Cookbook](https://github.com/jvns/pandas-cookbook) | Great resource with examples from weather, bikes and 311 calls
[Learn Pandas Tutorials](https://bitbucket.org/hrojas/learn-pandas) | A great series of Pandas tutorials from Dave Rojas
[Research Computing Python Data PYNBs](https://github.com/ResearchComputing/Meetup-Fall-2013/tree/master/python) | A super awesome set of python notebooks from a meetup-based course exclusively devoted to pandas